pkg/operator/controller/dns/controller.go

```diff
@@ -0,0 +1,157 @@
+
+	dnsConfig := &configv1.DNS{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: "cluster"}, dnsConfig); err != nil {
+		errs = append(errs, fmt.Errorf("failed to get dns 'cluster': %v", err))
+		return result, utilerrors.NewAggregate(errs)
+	}
+
+	record := &iov1.DNSRecord{}
+	if err := r.client.Get(context.TODO(), request.NamespacedName, record); err != nil {
+		if errors.IsNotFound(err) {
+			log.Info("dnsrecord not found; reconciliation will be skipped", "request", request)
+		} else {
+			errs = append(errs, fmt.Errorf("failed to get dnsrecord: %v", err))
+		}
+	}
+
+	ingressName, ok := record.Labels[manifests.OwningIngressControllerLabel]
+	if !ok {
+		errs = append(errs, fmt.Errorf("missing owner label"))
+		return result, utilerrors.NewAggregate(errs)
+	}
+
+	ingress := &operatorv1.IngressController{}
+	if err := r.cache.Get(context.TODO(), types.NamespacedName{Namespace: record.Namespace, Name: ingressName}, ingress); err != nil {
+		errs = append(errs, fmt.Errorf("failed to get dnsrecord for record %s: %v", record.Name, err))
+		return result, utilerrors.NewAggregate(errs)
+	}
+
+	// TODO: filter list based on igresscontroller LB scope
+	zones := []configv1.DNSZone{}
+	if dnsConfig.Spec.PrivateZone != nil {
+		zones = append(zones, *dnsConfig.Spec.PrivateZone)
+	}
+	if dnsConfig.Spec.PublicZone != nil {
+		zones = append(zones, *dnsConfig.Spec.PublicZone)
+	}
+
+	// If the DNS record was deleted, clean up and return.
+	if record.DeletionTimestamp != nil {
+		for i := range zones {
+			zone := zones[i]
+			err := r.dnsProvider.Delete(record, zone)
+			if err != nil {
+				errs = append(errs, err)
+			} else {
+				log.Info("provider deleted record %s in zone %v", record.Name, zone)
+			}
+		}
+		if len(errs) == 0 {
+			updated := record.DeepCopy()
+			if slice.ContainsString(updated.Finalizers, "operator.openshift.io/ingress") {
+				updated.Finalizers = slice.RemoveString(updated.Finalizers, "operator.openshift.io/ingress")
+				if err := r.client.Update(context.TODO(), updated); err != nil {
+					errs = append(errs, fmt.Errorf("failed to remove finalizer from dnsrecord %s: %v", record.Name, err))
+				}
+			}
+		}
+		return result, utilerrors.NewAggregate(errs)
+	}
+
+	statuses := []iov1.DNSZoneStatus{}
+	for i := range zones {
+		zone := zones[i]
+		conditions := []iov1.DNSZoneCondition{}
+		err := r.dnsProvider.Ensure(record, zone)
+		if err != nil {
+			conditions = append(conditions, iov1.DNSZoneCondition{
+				Type:               iov1.DNSRecordFailedConditionType,
+				Status:             string(operatorv1.ConditionTrue),
+				Reason:             "ProviderError",
+				Message:            fmt.Sprintf("The DNS provider failed to ensure the record: %v", err),
+				LastTransitionTime: metav1.Now(),
+			})
+			// TODO: return a retry result?
+		}
+		statuses = append(statuses, iov1.DNSZoneStatus{
+			DNSZone:    zone,
+			Conditions: conditions,
+		})
+	}
+
+	// TODO: only update if status changed
```

Comment:
Implement this and the AWS client caching can be deleted

Reply:
Improved the TODO but don't intend to fix now because:

1. It's a bit of work to get right
2. It involves refactoring the DNS provider implementations
3. We're already better off than before because we now only call ensure on creation (so once during startup, and once during ingresscontroller creation), so fixing the last problem here has low ROI for this particular PR (which is already huge)